datafusion/physical-expr/src/aggregate/mod.rs

```diff
@@ -102,6 +102,21 @@ pub trait AggregateExpr: Send + Sync + Debug + PartialEq<dyn Any> {
         "AggregateExpr: default name"
     }
 
+    /// Returns Aggregate Fucntion Name
+    fn func_name(&self) -> &str;
+
+    /// Human readable name such as `"MIN(c2)"` or `"RANK()"`. The default
+    /// implementation returns `"FUNCTION_NAME(args)"`
+    fn display_name(&self) -> String {
```

Comment:
I am confused by this API change -- the AggregateFunction now has `name()`, `func_name()` **and** `display_name()`

I thought the reason for having `name()` was so that that the code that created the AggregateExpr would calculate the name once and then pass it along. The creation of the name happened *once* on construction rather than on demand

For example:
https://github.com/apache/arrow-datafusion/blob/4c914ea1e5d3dc61f3552adcffb8356c90d73bac/datafusion/proto/src/physical_plan/mod.rs#L464-L470 

This PR seems to have added a parallel API for creating display names for aggregate functions which I think is quite confusing

Reply:
I will try to dynamic calculation from implementation

Reply:
I gave this a try. However, replacing name with current display name did not work. It seems like during schema creation aggregate exprs uses `.name` for the field (See [link](https://github.com/apache/arrow-datafusion/blob/06bbe1298fa8aa042b6a6462e55b2890969d884a/datafusion/physical-expr/src/aggregate/mod.rs#L77)). Hence plan depends on the `.name` string for successful execution. Otherwise subsequent projection after (window or aggregation) may refer to invalid fields. 

Other option might be adding a `display_name` member to the each aggregate expression (similar to `.name` method). This `display_name` member will not be set from argument but will calculate at the initialization(Users will not have control over it). Then we will use `.display_name(&self) -> &str` API to return this member. With this way `func_name` argument will be removed. However, we will still have parallel `.display_name` on top `.name` which might be confusing. However, I think this approach is better than the current one. 

What do you think?

Reply:
I was thinking that we could add a `name` field to `WindowFunction` upon creation to mirror what happens with AggregateFunction. That way the code would be consistent and we can use the same pattern that works for AggregateFunction

Reply:
The fact that we have `name`, `func_name` and `display_name` is an indication that we can do a better design. @mustafasrepo and I will discuss to see what we can do here.